flang/lib/Optimizer/Dialect/FIRType.cpp

```diff
@@ -918,11 +919,23 @@ bool fir::VectorType::isValidElementType(mlir::Type t) {
 //===----------------------------------------------------------------------===//
 // FIROpsDialect
 //===----------------------------------------------------------------------===//
-
+namespace {
+/// Model for FIR pointer like types that already provide a `getElementType`
+/// method
+template <typename T>
+struct PointerLikeModel
+    : public mlir::omp::PointerLikeType::ExternalModel<PointerLikeModel<T>, T> {
+  mlir::Type getElementType(mlir::Type pointer) const {
+    return pointer.cast<T>().getElementType();
+  }
+};
+} // end namespace
 void FIROpsDialect::registerTypes() {
   addTypes<BoxType, BoxCharType, BoxProcType, CharacterType, fir::ComplexType,
            FieldType, HeapType, fir::IntegerType, LenType, LogicalType,
            LLVMPointerType, PointerType, RealType, RecordType, ReferenceType,
            SequenceType, ShapeType, ShapeShiftType, ShiftType, SliceType,
            TypeDescType, fir::VectorType>();
+  fir::ReferenceType::attachInterface<PointerLikeModel<fir::ReferenceType>>(
```

Comment:
@schweitzpgi Are you OK with this change here?

Reply:
`ReferenceType` does behave like a pointer, so I think this is ok.

`ReferenceType` does not have C-like pointer semantics, of course. It is more refined, limited.

Is this sufficient? `HeapType`, `PointerType`, and `LLVMPointerType` also behave like pointers (with limitations on the first two types.

Reply:
`HeapType` and `LLVMPointerType` don't have a `getElementType` member. Hence adding their interface failed. `PointerType` I have added alongside `ReferenceType`

Reply:
@kiranchandramohan @schweitzpgi Would this be fine? Or do `HeapType` and `LLVMPointerType` have some other member that behaves like `getElementType`?

Reply:
HeapType and LLVMPointerType have a `getEleTy()` function.

Reply:
This worked actually. Thanks @shraiysh.